controlplane/kubeadm/controllers/helpers.go

```diff
@@ -246,3 +248,87 @@ func (r *KubeadmControlPlaneReconciler) generateMachine(ctx context.Context, kcp
 	}
 	return nil
 }
+
+// MachinesNeedingRollout return a list of machines that need to be upgraded.
+func (r *KubeadmControlPlaneReconciler) MachinesNeedingRollout(ctx context.Context, c *internal.ControlPlane) internal.FilterableMachineCollection {
+	machinesToUpgrade := internal.NewFilterableMachineCollection()
+
+	now := metav1.Now()
+	if c.KCP.Spec.UpgradeAfter != nil && c.KCP.Spec.UpgradeAfter.Before(&now) {
+		oldMachines := c.Machines.Filter(machinefilters.OlderThan(c.KCP.Spec.UpgradeAfter))
+		// Ignore deleted machines.
+		// TODO: This should be removed once MHC remediation is merged.
+		oldMachines = oldMachines.Filter(func(machine *clusterv1.Machine) bool {
+			return machine.DeletionTimestamp.IsZero()
+		})
+		for _, m := range oldMachines {
+			machinesToUpgrade.Insert(m)
+		}
+	}
+
+	for _, machine := range c.Machines {
+		// TODO: This should be removed once MHC remediation is merged.
+		if !machine.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		if *machine.Spec.Version != c.KCP.Spec.Version {
+			machinesToUpgrade.Insert(machine)
+			continue
+		}
+
+		ref := machine.Spec.Bootstrap.ConfigRef
+		if ref == nil {
+			continue
+		}
+
+		machineCfg := &bootstrapv1.KubeadmConfig{}
+		if err := r.Client.Get(ctx, client.ObjectKey{Name: ref.Name, Namespace: c.KCP.Namespace}, machineCfg); err != nil {
+			continue
+		}
+
+		kcpSpec := createKcpKubeadmConfig(machineCfg.Spec, c.KCP.Spec.KubeadmConfigSpec)
+
+		if !reflect.DeepEqual(machineCfg.Spec, kcpSpec) {
+			machinesToUpgrade.Insert(machine)
+			continue
+		}
+
+		infraObj, err := external.Get(ctx, r.Client, &machine.Spec.InfrastructureRef, machine.Spec.InfrastructureRef.Namespace)
+		if err != nil {
+			//
+			continue
+		}
+
+		clonedFromName, existName := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromNameAnnotation]
+		clonedFromGroupKind, existGroupKind := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromGroupKindAnnotation]
+		if !existName || !existGroupKind {
+			// All KCP cloned infra machines should have this annotation.
+			// Missing the annotation here may be due to older version machines or adopted machines.
+			continue
+		}
+
+		if clonedFromName != c.KCP.Spec.InfrastructureTemplate.Name || clonedFromGroupKind != c.KCP.Spec.InfrastructureTemplate.GroupVersionKind().GroupKind().String() {
+			machinesToUpgrade.Insert(machine)
+			continue
+		}
+	}
+
+	return machinesToUpgrade
+}
+
+// TODO: should we accept machines missing all 3 configs - init, join, and cluster?
```

Comment:
We should probably rollout the Machine if this is the case, KCP owned Machines should ad least have one between init and join

Reply:
In the KubeamConfig controller, we allow for defaulting an empty JoinConfiguration if one is not provided, we should likely do the same here as well.

Reply:
Okay, makes sense.

Reply:
Also only both init and cluster config is missing, KubeadmConfig controller is returning. I will follow the same logic here.

Reply:
Empty as in nil?

Reply:
Indeed: https://github.com/kubernetes-sigs/cluster-api/blob/241d24738094c617f1c686cea121f17d01df3c48/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go#L276-L280

Reply:
So to be clear, we will not allow all 3 configs to be empty because join or init will be defaulted.

Reply:
The only reason that we require the configs in KubeadmConfig when doing an init is because we needed to differentiate if we should be doing an init vs a join. In our case, With KCP, we could potentially default to an empty config if not provided since we have the context to determine when to init vs when to join.

As an aside, we don't actually need InitConfiguration after the cluster already exists, and only really need ClusterConfiguration after the cluster already exists if we are managing the overrides provided by it for a user, so we really shouldn't need either of those in the case of adoption.

Reply:
Combined with the comment below, I think only check here should be:

if InitConfiguration == nil && JoinConfiguration == nil && ClusterConfiguration == nil {
    return err
}

Having nil ClusterConfiguration or InitConfiguration is allowed when controlplane is not initialized.